File: src/test/kotlin/no/nav/bidrag/arbeidsflyt/hendelse/JournalpostHendelseListenerTest.java
package no.nav.bidrag.arbeidsflyt.hendelse;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.Mockito.verify;

import com.fasterxml.jackson.databind.ObjectMapper;
import no.nav.bidrag.arbeidsflyt.service.HendelseService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class JournalpostHendelseListenerTest {

  private static final String HENDELSE_TEMPLATE = """
      {
        "journalpostId":"BID-%s",
        "hendelse":"%s",
        "sporing": {
          "correlationId":"xyz",
          "opprettet":"nå"
        }
      }
      """.stripIndent().trim();

  @Mock
  private HendelseService hendelseServiceMock;
  private JournalpostHendelseListener journalpostHendelseListener;

  @BeforeEach
  void initTestClass() {
    journalpostHendelseListener = new JournalpostHendelseListener(new ObjectMapper(), hendelseServiceMock);
  }

  @Test
  @DisplayName("skal behandle journalpost hendelse")
  void skalBehandleJournalpostHendelse() {
    journalpostHendelseListener.lesHendelse(String.format(HENDELSE_TEMPLATE, 1, "TEST_HENDELSE"));

    verify(hendelseServiceMock).behandleHendelse(new JournalpostHendelse("BID-1", "TEST_HENDELSE"));
  }

  @Test
  @DisplayName("skal ha sporingsdata på en hendelse")
  void skalHaSporingsdataPaHendelse() {
    journalpostHendelseListener.lesHendelse(String.format(HENDELSE_TEMPLATE, 2, "TEST_HENDELSE"));

    var argumentCaptor = ArgumentCaptor.forClass(JournalpostHendelse.class);

    verify(hendelseServiceMock).behandleHendelse(argumentCaptor.capture());

    assertThat(argumentCaptor.getValue().getSporing()).isEqualTo(new Sporingsdata("xyz", "nå"));
  }
}

File: src/test/kotlin/no/nav/bidrag/arbeidsflyt/hendelse/JournalpostHendelseListenerTest.kt
package no.nav.bidrag.arbeidsflyt.hendelse

import no.nav.bidrag.arbeidsflyt.service.BehandleHendelseService
import no.nav.bidrag.dokument.dto.JournalpostHendelse
import no.nav.bidrag.dokument.dto.Sporingsdata
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito.verify
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.context.ActiveProfiles

@SpringBootTest
@DisplayName("JournalpostHendelseListener")
@ActiveProfiles("test")
internal class JournalpostHendelseListenerTest {

    @Autowired
    private lateinit var journalpostHendelseListener: JournalpostHendelseListener

    @MockBean
    private lateinit var behandleHendelseServiceMock: BehandleHendelseService

    @Test
    fun `skal mappe og behandle JournalpostHendelse`() {
        journalpostHendelseListener.lesHendelse(
            """{
              "journalpostId":"BID-1"
            }""".trimIndent()
        )

        verify(behandleHendelseServiceMock).behandleHendelse(JournalpostHendelse(journalpostId = "BID-1"))
    }

    @Test
    fun `skal ha sporingsdata i meldingen`() {
        journalpostHendelseListener.lesHendelse(
            """{
              "journalpostId":"BID-2",
              "sporing": {
                "correlationId":"xyz",
                "brukerident":"jb",
                "saksbehandlersNavn":"Jon Blund"
              }
            }""".trimIndent()
        )

        verify(behandleHendelseServiceMock).behandleHendelse(
            JournalpostHendelse(
                journalpostId = "BID-2",
                sporing = Sporingsdata(
                    correlationId = "xyz",
                    brukerident = "jb",
                    saksbehandlersNavn = "Jon Blund"
                )
            )
        )
    }
}
